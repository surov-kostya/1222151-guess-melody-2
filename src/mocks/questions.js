import {shape, string, oneOf, arrayOf, number} from 'prop-types';

export const questionType = shape({
  type: oneOf([`genre`, `artist`]).isRequired,
  genrge: oneOf([`rock`, `folk`, `jazz`, `pop`]),
  song: shape({
    src: string,
    artist: string
  }),
  answers: arrayOf(
      shape({
        id: number,
        src: string.isRequired,
        genrge: oneOf([`rock`, `folk`, `jazz`, `pop`]).isRequired
      })) |
    shape({
      id: number,
      picture: string.isRequired,
      artist: string.isRequired
    }).isRequired
});

export const settings = {
  gameTime: 5,
  errorCount: 3,
};

export const gameQuestions = [
  {
    type: `genre`,
    genre: `Hip Hop & Rap`,
    answers: [
      {
        id: 100,
        src: `https://www.youtube.com/audiolibrary_download?vid=3b32fa1674bc3764`,
        genre: `Hip Hop & Rap`,
      },
      {
        id: 101,
        src: `https://www.youtube.com/audiolibrary_download?vid=28782f0fee64fd45`,
        genre: `Dance & Electronic`,
      },
      {
        id: 102,
        src: `https://www.youtube.com/audiolibrary_download?vid=1b61f851e58e6ffe`,
        genre: `pop`,
      },
      {
        id: 103,
        src: `https://www.youtube.com/audiolibrary_download?vid=ad2e6ac237dfb8ba`,
        genre: `Dance & Electronic`,
      },
    ],
  },
  {
    type: `genre`,
    genre: `Pop`,
    answers: [
      {
        id: 104,
        src: `https://www.youtube.com/audiolibrary_download?vid=1b61f851e58e6ffe`,
        genre: `Pop`,
      },
      {
        id: 105,
        src: `https://www.youtube.com/audiolibrary_download?vid=ad2e6ac237dfb8ba`,
        genre: `Dance & Electronic`,
      },
      {
        id: 106,
        src: `https://www.youtube.com/audiolibrary_download?vid=47be4e2187adf0c5`,
        genre: `Dance & Electronic`,
      },
      {
        id: 107,
        src: `https://www.youtube.com/audiolibrary_download?vid=fdce0df3f10e0a6f`,
        genre: `Ambient`,
      },
    ],
  },
  {
    type: `artist`,
    song: {
      artist: `Bad Snacks`,
      src: `https://www.youtube.com/audiolibrary_download?vid=47be4e2187adf0c5`,
    },
    answers: [
      {
        id: 200,
        picture: `http://placehold.it/134x134`,
        artist: `Bad Snacks`,
      },
      {
        id: 201,
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        id: 202,
        picture: `http://placehold.it/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
  {
    type: `artist`,
    song: {
      artist: `Jeremy Blake`,
      src: `https://www.youtube.com/audiolibrary_download?vid=3b32fa1674bc3764`,
    },
    answers: [
      {
        id: 203,
        picture: `http://placehold.it/134x134`,
        artist: `Jeremy Blake`,
      },
      {
        id: 204,
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        id: 205,
        picture: `http://placehold.it/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
];

export const songs = [
  {
    "title": `U In My Arms`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=3b32fa1674bc3764`,
    "artist": `Jeremy Blake`,
    "genre": `Hip Hop & Rap`,
  },
  {
    "title": `Through The Crystal`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=28782f0fee64fd45`,
    "artist": `Jeremy Blake`,
    "genre": `Dance & Electronic`,
  },
  {
    "title": `Everything Is Gonna Be Just Fine`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=1b61f851e58e6ffe`,
    "artist": `Jeremy Blake`,
    "genre": `Pop`,
  },
  {
    "title": `Block Party`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=ad2e6ac237dfb8ba`,
    "artist": `Bad Snacks`,
    "genre": `Dance & Electronic`,
  },
  {
    "title": `New Moon`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=47be4e2187adf0c5`,
    "artist": `Bad Snacks`,
    "genre": `Dance & Electronic`,
  },
  {
    "title": `They Might Not`,
    "download_url": `https://www.youtube.com/audiolibrary_download?vid=fdce0df3f10e0a6f`,
    "artist": `Puddle of Infinity`,
    "genre": `Ambient`,
  }
];
