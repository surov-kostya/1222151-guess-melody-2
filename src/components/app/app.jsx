import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {questionType} from '../../mocks/questions';

import {ActionCreator} from '../../reducer';
import Welcome from '../welcome/welcome';
import GameArtist from '../game-artist/game-artist';
import GameGenre from '../game-genre/game-genre';
import MistakesCounter from '../mistakes-counter/mistakes-counter';
import Timer from '../timer/timer';
import GameOver from '../game-over/game-over';

class App extends PureComponent {
  static getScreen(props) {
    const {
      gameTime,
      errorCount,
      onWelcomeScreenClick,
      step,
      questions,
      isGameOver,
      onRestart
    } = props;

    if (isGameOver) {
      return (
        <GameOver onRestart={onRestart} />
      );
    }

    if (step !== -1) {
      const {mistakes, onUserAnswer, onTimeout, onTick} = props;
      const currentQuestion = questions[step];
      const stepsToEnd = questions.length - step - 1;

      switch (currentQuestion.type) {
        case `genre`: return <GameGenre
          question={currentQuestion}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, errorCount, stepsToEnd)}
        >
          <Timer time={gameTime} onTimeout={onTimeout} onTick={(time) => onTick(time)}/>
          <MistakesCounter mistakes={mistakes}/>
        </GameGenre>;

        case `artist`: return <GameArtist
          screenIndex={step}
          question={currentQuestion}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, errorCount, stepsToEnd)}
        >
          <Timer time={gameTime} onTimeout={onTimeout} onTick={(time) => onTick(time)}/>
          <MistakesCounter mistakes={mistakes}/>
        </GameArtist>;
      }
    }

    return <Welcome
      gameTime={gameTime}
      availableMistakes={errorCount}
      onStart={onWelcomeScreenClick}
    />;
  }

  constructor(props) {
    super(props);

    this.state = {
      step: -1,
      answers: []
    };
  }

  render() {
    return App.getScreen(this.props);
  }
}

App.propTypes = {
  questions: PropTypes.arrayOf(questionType),
  gameTime: PropTypes.number,
  errorCount: PropTypes.number,
  onWelcomeScreenClick: PropTypes.func.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
  onTimeout: PropTypes.func.isRequired,
  onTick: PropTypes.func.isRequired,
  onRestart: PropTypes.func.isRequired,
  step: PropTypes.number.isRequired,
  mistakes: PropTypes.number.isRequired,
  isGameOver: PropTypes.bool.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  step: state.step,
  mistakes: state.mistakes,
  gameTime: state.gameTime,
  isGameOver: state.isGameOver
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incStep()),
  onUserAnswer: (userAnswer, question, mistakes, maxMistakes, stepsToEnd) => {
    dispatch(ActionCreator.incStep());
    dispatch(ActionCreator.incMistakes(userAnswer, question, mistakes, maxMistakes, stepsToEnd));
  },
  onTimeout: () => dispatch(ActionCreator.gameOver()),
  onTick: (time) => dispatch(ActionCreator.decGameTime(time)),
  onRestart: () => dispatch(ActionCreator.reset())
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
