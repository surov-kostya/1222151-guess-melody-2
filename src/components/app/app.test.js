import React from 'react';
import {App} from './app';
import renderer from 'react-test-renderer';
import {gameQuestions} from '../../mocks/questions';

it(`App renders correctly`, () => {
  const tree = renderer
    .create(<App questions={gameQuestions} gameTime={7} errorCount={3}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
