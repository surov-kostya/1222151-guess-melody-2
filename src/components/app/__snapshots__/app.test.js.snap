// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        0
        0
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        07
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          style={
            Object {
              "outline": "none",
            }
          }
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={203}
          name="answer"
          onChange={[Function]}
          type="radio"
          value={203}
        />
        <label
          className="artist__name"
          htmlFor={203}
        >
          <img
            alt="Jeremy Blake"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          Jeremy Blake
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={204}
          name="answer"
          onChange={[Function]}
          type="radio"
          value={204}
        />
        <label
          className="artist__name"
          htmlFor={204}
        >
          <img
            alt="Jack Daniels"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          Jack Daniels
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={205}
          name="answer"
          onChange={[Function]}
          type="radio"
          value={205}
        />
        <label
          className="artist__name"
          htmlFor={205}
        >
          <img
            alt="Jim Beam"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          Jim Beam
        </label>
      </div>
    </form>
  </section>
</section>
`;
